File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/core/src/types/VLMTypes.ts
/**
 * VLMTypes.ts
 * Type definitions for Vision Language Model (VLM) functionality.
 * Reference: sdk/runanywhere-swift/Sources/RunAnywhere/Public/Extensions/VLM/VLMTypes.swift
 */

/** VLM image format enum - matches rac_vlm_image_format_t */
export enum VLMImageFormat {
  FilePath = 0,
  RGBPixels = 1,
  Base64 = 2,
}

/** VLM image input - discriminated union matching Swift VLMImage.Format */
export type VLMImage =
  | { format: VLMImageFormat.FilePath; filePath: string }
  | { format: VLMImageFormat.RGBPixels; data: Uint8Array; width: number; height: number }
  | { format: VLMImageFormat.Base64; base64: string };

/** VLM generation options */
export interface VLMGenerationOptions {
  maxTokens?: number;       // default 2048
  temperature?: number;     // default 0.7
  topP?: number;           // default 0.9
}

/** VLM generation result - matches Swift VLMResult */
export interface VLMResult {
  text: string;
  promptTokens: number;
  completionTokens: number;
  totalTimeMs: number;
  tokensPerSecond: number;
}

/** VLM streaming result - matches Swift VLMStreamingResult */
export interface VLMStreamingResult {
  stream: AsyncIterable<string>;
  result: Promise<VLMResult>;
  cancel: () => void;
}

/** VLM error codes - matches Swift SDKError.VLMErrorCode */
export enum VLMErrorCode {
  NotInitialized = 1,
  ModelLoadFailed = 2,
  ProcessingFailed = 3,
  InvalidImage = 4,
  Cancelled = 5,
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-flutter/packages/runanywhere/lib/public/types/vlm_types.dart
/// VLM (Vision Language Model) Types
///
/// Public types for VLM image processing.
/// Mirrors iOS VLMTypes.swift adapted for Flutter/Dart.
library vlm_types;

import 'dart:typed_data';

// MARK: - VLM Image Input (Dart-adapted)

/// Image input for VLM - handles Dart-native image formats
///
/// Supports three image formats:
/// - filePath: Path to an image file (JPEG, PNG, etc.)
/// - rgbPixels: Raw RGB pixel data (RGBRGBRGB layout)
/// - base64: Base64-encoded image data
///
/// Matches iOS VLMImage but uses Dart-native types instead of UIImage/CVPixelBuffer.
class VLMImage {
  final VLMImageFormat format;

  const VLMImage._(this.format);

  /// Create from a file path (JPEG, PNG, etc.)
  factory VLMImage.filePath(String path) => VLMImage._(VLMImageFormat.filePath(path));

  /// Create from raw RGB pixel data (RGBRGBRGB layout)
  factory VLMImage.rgbPixels(Uint8List data, {required int width, required int height}) =>
      VLMImage._(VLMImageFormat.rgbPixels(data: data, width: width, height: height));

  /// Create from base64-encoded image data
  factory VLMImage.base64(String encoded) => VLMImage._(VLMImageFormat.base64(encoded));
}

/// Image format variants (sealed class for type safety)
sealed class VLMImageFormat {
  const VLMImageFormat();

  factory VLMImageFormat.filePath(String path) = VLMImageFormatFilePath;
  factory VLMImageFormat.rgbPixels({required Uint8List data, required int width, required int height}) = VLMImageFormatRgbPixels;
  factory VLMImageFormat.base64(String encoded) = VLMImageFormatBase64;
}

/// File path format
class VLMImageFormatFilePath extends VLMImageFormat {
  final String path;
  const VLMImageFormatFilePath(this.path);
}

/// RGB pixels format
class VLMImageFormatRgbPixels extends VLMImageFormat {
  final Uint8List data;
  final int width;
  final int height;
  const VLMImageFormatRgbPixels({required this.data, required this.width, required this.height});
}

/// Base64 format
class VLMImageFormatBase64 extends VLMImageFormat {
  final String encoded;
  const VLMImageFormatBase64(this.encoded);
}

// MARK: - VLM Result

/// Result from VLM generation
/// Matches iOS VLMResult
class VLMResult {
  /// Generated text describing the image
  final String text;

  /// Number of tokens in the prompt (including image tokens)
  final int promptTokens;

  /// Number of tokens generated in the response
  final int completionTokens;

  /// Total processing time in milliseconds
  final double totalTimeMs;

  /// Tokens generated per second
  final double tokensPerSecond;

  const VLMResult({
    required this.text,
    required this.promptTokens,
    required this.completionTokens,
    required this.totalTimeMs,
    required this.tokensPerSecond,
  });

  @override
  String toString() {
    final textPreview = text.length > 50 ? text.substring(0, 50) : text;
    return 'VLMResult(text: "$textPreview...", tokens: $completionTokens, ${tokensPerSecond.toStringAsFixed(1)} tok/s)';
  }
}

// MARK: - VLM Streaming

/// Streaming result for VLM generation
/// Matches iOS VLMStreamingResult, adapted for Dart async patterns
class VLMStreamingResult {
  /// Stream of tokens as they are generated
  final Stream<String> stream;

  /// Future that completes with final result metrics when streaming finishes
  final Future<VLMResult> metrics;

  /// Function to cancel the ongoing generation
  final void Function() cancel;

  const VLMStreamingResult({
    required this.stream,
    required this.metrics,
    required this.cancel,
  });
}

// MARK: - VLM Error Codes

/// VLM-specific error codes
/// Matches iOS SDKError.VLMErrorCode exactly
enum VLMErrorCode {
  /// VLM model not loaded
  notInitialized(1),

  /// Model load operation failed
  modelLoadFailed(2),

  /// Image processing failed
  processingFailed(3),

  /// Invalid image input
  invalidImage(4),

  /// Generation was cancelled
  cancelled(5);

  final int rawValue;
  const VLMErrorCode(this.rawValue);
}
